test/data/iris_decimal_marker_as_comma.txt

```diff
@@ -0,0 +1,151 @@
+#Label	Sepal length	Sepal width	Petal length	Petal width
+0	5,1	3,5	1,4	0,2
+0	4,9	3,0	1,4	0,2
+0	4,7	3,2	1,3	0,2
+0	4,6	3,1	1,5	0,2
+0	5,0	3,6	1,4	0,2
+0	5,4	3,9	1,7	0,4
+0	4,6	3,4	1,4	0,3
+0	5,0	3,4	1,5	0,2
+0	4,4	2,9	1,4	0,2
+0	4,9	3,1	1,5	0,1
+0	5,4	3,7	1,5	0,2
+0	4,8	3,4	1,6	0,2
+0	4,8	3,0	1,4	0,1
+0	4,3	3,0	1,1	0,1
+0	5,8	4,0	1,2	0,2
+0	5,7	4,4	1,5	0,4
+0	5,4	3,9	1,3	0,4
+0	5,1	3,5	1,4	0,3
+0	5,7	3,8	1,7	0,3
+0	5,1	3,8	1,5	0,3
+0	5,4	3,4	1,7	0,2
+0	5,1	3,7	1,5	0,4
+0	4,6	3,6	1,0	0,2
+0	5,1	3,3	1,7	0,5
+0	4,8	3,4	1,9	0,2
+0	5,0	3,0	1,6	0,2
+0	5,0	3,4	1,6	0,4
+0	5,2	3,5	1,5	0,2
+0	5,2	3,4	1,4	0,2
+0	4,7	3,2	1,6	0,2
+0	4,8	3,1	1,6	0,2
+0	5,4	3,4	1,5	0,4
+0	5,2	4,1	1,5	0,1
+0	5,5	4,2	1,4	0,2
+0	4,9	3,1	1,5	0,1
+0	5,0	3,2	1,2	0,2
+0	5,5	3,5	1,3	0,2
+0	4,9	3,1	1,5	0,1
+0	4,4	3,0	1,3	0,2
+0	5,1	3,4	1,5	0,2
+0	5,0	3,5	1,3	0,3
+0	4,5	2,3	1,3	0,3
+0	4,4	3,2	1,3	0,2
+0	5,0	3,5	1,6	0,6
+0	5,1	3,8	1,9	0,4
+0	4,8	3,0	1,4	0,3
+0	5,1	3,8	1,6	0,2
+0	4,6	3,2	1,4	0,2
+0	5,3	3,7	1,5	0,2
+0	5,0	3,3	1,4	0,2
+1	7,0	3,2	4,7	1,4
+1	6,4	3,2	4,5	1,5
+1	6,9	3,1	4,9	1,5
+1	5,5	2,3	4,0	1,3
+1	6,5	2,8	4,6	1,5
+1	5,7	2,8	4,5	1,3
+1	6,3	3,3	4,7	1,6
+1	4,9	2,4	3,3	1,0
+1	6,6	2,9	4,6	1,3
+1	5,2	2,7	3,9	1,4
+1	5,0	2,0	3,5	1,0
+1	5,9	3,0	4,2	1,5
+1	6,0	2,2	4,0	1,0
+1	6,1	2,9	4,7	1,4
+1	5,6	2,9	3,6	1,3
+1	6,7	3,1	4,4	1,4
+1	5,6	3,0	4,5	1,5
+1	5,8	2,7	4,1	1,0
+1	6,2	2,2	4,5	1,5
+1	5,6	2,5	3,9	1,1
+1	5,9	3,2	4,8	1,8
+1	6,1	2,8	4,0	1,3
+1	6,3	2,5	4,9	1,5
+1	6,1	2,8	4,7	1,2
+1	6,4	2,9	4,3	1,3
+1	6,6	3,0	4,4	1,4
+1	6,8	2,8	4,8	1,4
+1	6,7	3,0	5,0	1,7
+1	6,0	2,9	4,5	1,5
+1	5,7	2,6	3,5	1,0
+1	5,5	2,4	3,8	1,1
+1	5,5	2,4	3,7	1,0
+1	5,8	2,7	3,9	1,2
+1	6,0	2,7	5,1	1,6
+1	5,4	3,0	4,5	1,5
+1	6,0	3,4	4,5	1,6
+1	6,7	3,1	4,7	1,5
+1	6,3	2,3	4,4	1,3
+1	5,6	3,0	4,1	1,3
+1	5,5	2,5	4,0	1,3
+1	5,5	2,6	4,4	1,2
+1	6,1	3,0	4,6	1,4
+1	5,8	2,6	4,0	1,2
+1	5,0	2,3	3,3	1,0
+1	5,6	2,7	4,2	1,3
+1	5,7	3,0	4,2	1,2
+1	5,7	2,9	4,2	1,3
+1	6,2	2,9	4,3	1,3
+1	5,1	2,5	3,0	1,1
+1	5,7	2,8	4,1	1,3
+2	6,3	3,3	6,0	2,5
+2	5,8	2,7	5,1	1,9
+2	7,1	3,0	5,9	2,1
+2	6,3	2,9	5,6	1,8
+2	6,5	3,0	5,8	2,2
+2	7,6	3,0	6,6	2,1
+2	4,9	2,5	4,5	1,7
+2	7,3	2,9	6,3	1,8
+2	6,7	2,5	5,8	1,8
+2	7,2	3,6	6,1	2,5
+2	6,5	3,2	5,1	2,0
+2	6,4	2,7	5,3	1,9
+2	6,8	3,0	5,5	2,1
+2	5,7	2,5	5,0	2,0
+2	5,8	2,8	5,1	2,4
+2	6,4	3,2	5,3	2,3
+2	6,5	3,0	5,5	1,8
+2	7,7	3,8	6,7	2,2
+2	7,7	2,6	6,9	2,3
+2	6,0	2,2	5,0	1,5
+2	6,9	3,2	5,7	2,3
+2	5,6	2,8	4,9	2,0
+2	7,7	2,8	6,7	2,0
+2	6,3	2,7	4,9	1,8
+2	6,7	3,3	5,7	2,1
+2	7,2	3,2	6,0	1,8
+2	6,2	2,8	4,8	1,8
+2	6,1	3,0	4,9	1,8
+2	6,4	2,8	5,6	2,1
+2	7,2	3,0	5,8	1,6
+2	7,4	2,8	6,1	1,9
+2	7,9	3,8	6,4	2,0
+2	6,4	2,8	5,6	2,2
+2	6,3	2,8	5,1	1,5
+2	6,1	2,6	5,6	1,4
+2	7,7	3,0	6,1	2,3
+2	6,3	3,4	5,6	2,4
+2	6,4	3,1	5,5	1,8
+2	6,0	3,0	4,8	1,8
+2	6,9	3,1	5,4	2,1
+2	6,7	3,1	5,6	2,4
+2	6,9	3,1	5,1	2,3
+2	5,8	2,7	5,1	1,9
+2	6,8	3,2	5,9	2,3
+2	6,7	3,3	5,7	2,5
+2	6,7	3,0	5,2	2,3
+2	6,3	2,5	5,0	1,9
+2	6,5	3,0	5,2	2,0
+2	6,2	3,4	5,4	2,3
+2	5,9	3,0	5,1	1,8
```

Comment:
For a CSV file, how does a user normally insert a comma as a [decimal separator](https://en.wikipedia.org/wiki/Decimal_separator)?

Do they quote it? Does the text loader allow quoted floats?

Reply:
That is a good point @justinormont , though currently the code does not allow for both decimal marker and separators to be the same character. 

https://github.com/dotnet/machinelearning/blob/a663f210db833e1d48654c86ce0ff753230d9447/src/Microsoft.ML.Data/DataLoadSave/Text/TextLoader.cs#L1222-L1225

As such this dataset would need to be edited so that the decimal marker is a `.` instead.

Reply:
_Un-resolving this thread because I have some thoughts on this_

So I wonder if we should actually use `,` as both a separator and a decimal separator precisely for this case.

The only valid way for a user to use `,` as a decimal separator on a CSV file is to actually quote the field as Justin did on his example. My understanding solely from looking the code, is that (even without this PR) the TextLoader shouldn't have any problems loading this: (i.e., it can load quoted float columns)

```
Label,Sepal,length,Sepal,width,Petal length,Petal width
 0,"5.1","3.5","1.4","0.2"
 0,"4.9","3.0","1.4","0.2"
 0,"4.7","3.2","1.3","0.2"
```

So, I haven't tested that, and it's just my impression for looking the code. But If that is allowed, then I think we should also allow the case of having `,`s as decimal separators for this case.
I believe the only change needed would be to remove any exception that's thrown for this particular case. And adding a test to show this case works would be useful.

---
In reply to: [428340558](https://github.com/dotnet/machinelearning/pull/5145#discussion_r428340558) [](ancestors = 428340558)

Reply:
Thanks Antonio. @justinormont I've added a version of our iris dataset in .csv format, with quotes around doubles, and separators and decimal markers are ','. This is for this test case I wrote, which confirms that this dataset can be loaded. Thank you for bringing this possibility to light.

https://github.com/dotnet/machinelearning/blob/6837f64f2325999a74aadbbd74cc670c789026cb/test/Microsoft.ML.Tests/TextLoaderTests.cs#L1060-L1137